File: nhglib-core/src/io/github/movementspeed/nhglib/graphics/shaders/attributes/PbrColorAttribute.java
package io.github.movementspeed.nhglib.graphics.shaders.attributes;

import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g3d.Attribute;
import com.badlogic.gdx.utils.GdxRuntimeException;

public class PbrColorAttribute extends Attribute {
    public final static String AlbedoColorAlias = "albedoColor";
    public final static String MetalnessValueAlias = "metalnessValue";
    public final static String RoughnessValueAlias = "roughnessValue";

    public final static long AlbedoColor = register(AlbedoColorAlias);
    public final static long MetalnessValue = register(MetalnessValueAlias);
    public final static long RoughnessValue = register(RoughnessValueAlias);

    protected static long Mask = AlbedoColor | MetalnessValue | RoughnessValue;

    public final static boolean is (final long mask) {
        return (mask & Mask) != 0;
    }

    public final static PbrColorAttribute createAlbedo(final Color color) {
        return new PbrColorAttribute(AlbedoColor, color);
    }

    public final static PbrColorAttribute createMetalness(final float metalness) {
        return new PbrColorAttribute(MetalnessValue, new Color(metalness, metalness, metalness, 1.0f));
    }

    public final static PbrColorAttribute createRoughness(final float roughness) {
        return new PbrColorAttribute(RoughnessValue, new Color(roughness, roughness, roughness, 1.0f));
    }

    public final Color color = new Color();

    public PbrColorAttribute (final long type) {
        super(type);
        if (!is(type)) throw new GdxRuntimeException("Invalid type specified");
    }

    public PbrColorAttribute (final long type, final Color color) {
        this(type);
        if (color != null) this.color.set(color);
    }

    public PbrColorAttribute (final long type, float r, float g, float b, float a) {
        this(type);
        this.color.set(r, g, b, a);
    }

    public PbrColorAttribute (final PbrColorAttribute copyFrom) {
        this(copyFrom.type, copyFrom.color);
    }

    @Override
    public Attribute copy () {
        return new PbrColorAttribute(this);
    }

    @Override
    public int hashCode () {
        int result = super.hashCode();
        result = 953 * result + color.toIntBits();
        return result;
    }

    @Override
    public int compareTo (Attribute o) {
        if (type != o.type) return (int)(type - o.type);
        return ((PbrColorAttribute)o).color.toIntBits() - color.toIntBits();
    }
}

File: nhglib-core/src/io/github/movementspeed/nhglib/graphics/shaders/attributes/PBRColorAttribute.kt
package io.github.movementspeed.nhglib.graphics.shaders.attributes

import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g3d.Attribute
import com.badlogic.gdx.utils.GdxRuntimeException

class PBRColorAttribute(type: Long) : Attribute(type) {

    val color = Color()

    init {
        if (!`is`(type)) throw GdxRuntimeException("Invalid type specified")
    }

    constructor(type: Long, color: Color?) : this(type) {
        if (color != null) this.color.set(color)
    }

    constructor(type: Long, r: Float, g: Float, b: Float, a: Float) : this(type) {
        this.color.set(r, g, b, a)
    }

    constructor(copyFrom: PBRColorAttribute) : this(copyFrom.type, copyFrom.color) {}

    override fun copy(): Attribute {
        return PBRColorAttribute(this)
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 953 * result + color.toIntBits()
        return result
    }

    override fun compareTo(o: Attribute): Int {
        return if (type != o.type) (type - o.type).toInt() else (o as PBRColorAttribute).color.toIntBits() - color.toIntBits()
    }

    companion object {
        val AlbedoColorAlias = "albedoColor"
        val MetalnessValueAlias = "metalnessValue"
        val RoughnessValueAlias = "roughnessValue"

        val AlbedoColor = Attribute.register(AlbedoColorAlias)
        val MetalnessValue = Attribute.register(MetalnessValueAlias)
        val RoughnessValue = Attribute.register(RoughnessValueAlias)

        protected var Mask = AlbedoColor or MetalnessValue or RoughnessValue

        fun `is`(mask: Long): Boolean {
            return mask and Mask != 0L
        }

        fun createAlbedo(color: Color): PBRColorAttribute {
            return PBRColorAttribute(AlbedoColor, color)
        }

        fun createMetalness(metalness: Float): PBRColorAttribute {
            return PBRColorAttribute(MetalnessValue, Color(metalness, metalness, metalness, 1.0f))
        }

        fun createRoughness(roughness: Float): PBRColorAttribute {
            return PBRColorAttribute(RoughnessValue, Color(roughness, roughness, roughness, 1.0f))
        }
    }
}
